Assets/MixedRealityToolkit-Preview/CameraCapture/CameraCaptureARFoundation.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#if USE_ARFOUNDATION
+
+using System;
+using Unity.Collections;
+using Unity.Collections.LowLevel.Unsafe;
+using UnityEngine;
+using UnityEngine.XR.ARExtensions;
+using UnityEngine.XR.ARFoundation;
+
+namespace Microsoft.MixedReality.Toolkit.Preview.CameraCapture
+{
+	public class CameraCaptureARFoundation : ICameraCapture
+	{
+		private CameraResolution resolution;
+		private Texture2D        captureTex;
+		private bool             ready = false;
+
+		/// <summary>
+		/// Is the camera completely initialized and ready to begin taking pictures?
+		/// </summary>
+		public bool  IsReady           { get { return ready; } }
+		/// <summary>
+		/// Is the camera currently already busy with taking a picture?
+		/// </summary>
+		public bool  IsRequestingImage { get { return false;  } }
+		/// <summary>
+		/// Field of View of the camera in degrees. This value is never ready until after 
+		/// initialization, and in many cases, isn't accurate until after a picture has
+		/// been taken. It's best to check this after each picture if you need it.
+		/// </summary>
+		public float FieldOfView       { get { 
+			Matrix4x4 proj = Matrix4x4.identity;
+			ARSubsystemManager.cameraSubsystem.TryGetProjectionMatrix(ref proj);
+			float     fov  = Mathf.Atan(1.0f / proj[1,1] ) * 2.0f * Mathf.Rad2Deg;
+			return fov;
+		} }
+
+		/// <summary>
+		/// Starts up and selects a device's camera, and finds appropriate picture settings
+		/// based on the provided resolution! 
+		/// </summary>
+		/// <param name="preferGPUTexture">Do you prefer GPU textures, or do you prefer a NativeArray of colors? Certain optimizations may be present to take advantage of this preference.</param>
+		/// <param name="resolution">Preferred resolution for taking pictures, note that resolutions are not guaranteed! Refer to CameraResolution for details.</param>
+		/// <param name="onInitialized">When the camera is initialized, this callback is called! Some cameras may return immediately, others may take a while. Can be null.</param>
+		public void Initialize(bool aPreferGPUTexture, CameraResolution aResolution, Action aOnInitialized)
+		{
+			resolution = aResolution;
+			
+			Action<ARSystemStateChangedEventArgs> handler = null;
+			handler = (aState) =>
+			{
+				// Camera and orientation data aren't ready until ARFoundation is actually tracking!
+				if (aState.state == ARSystemState.SessionTracking)
+				{
+					ARSubsystemManager.systemStateChanged -= handler;
+					ready = true;
+					if (aOnInitialized != null)
+						aOnInitialized();
+				}
+			};
+			ARSubsystemManager.systemStateChanged += handler;
+		}
+		
+		private unsafe void GrabScreen()
```

Comment:
Can this be done without unsafe? This puts a requirement to compile the entire project as unsafe.

Reply:
Seems like there was a safe method hidden deep in the docs too, I switched over to that!